py-polars/polars/convert.py

```diff
@@ -133,7 +133,7 @@ def from_arrow(
     if isinstance(a, pa.Table):
         return pl.DataFrame.from_arrow(a, rechunk)
     elif isinstance(a, pa.Array):
```

Comment:
Shouldn't we keep utilize the same tactic as with the DataFrame? Dispatching to a private `Series._from_arrow`. This can clean up the Series constructor and we already know we have an arrow array, so that checks don't have to be done anymore (which are done in the generic constructor).

Reply:
It's a bit different from `DataFrame` right now, because `DataFrame` doesn't support `arrow` data yet in its constructor.

In any case, I'm not sold on the `DataFrame._from_x` methods. They just call the class constructor, which means they do the checks anyway. Those checks are really cheap so I don't see a problem with that. But it's cleaner to just delete the `DataFrame._from_x` method and call the constructor directly from `pl.from_x`.

Reply:
IMO it should be the other way around. :thinking: 

So the class constructor check the types and dispatches to the specialized `DataFrame._from_x`. That way the class constructor can have lean `if else tree` and the specialized constructors have their specialized logic.

```
__init__():

    if foo:
       self._from_foo(..)
    else if bar:
      self._from_bar(..)
```

Reply:
As an example, look at `DataFrame._from_dict`. It's kind of ridiculous; we're much better off calling the constructor from `pl.from_dict`.

Reply:
Yes, in that case it is. That's why I think the lines should be swapped. 😁

Reply:
I get where you're coming from, but it doesn't work nicely because:
* We want `DataFrame._from_x` to return a new `DataFrame` object
* In the `__init__` function,  we just want to instantiate the `DataFrame` object we are already building

So we cannot dispatch from the `__init__` to a `_from_x` classmethod, because it doesn't give us what we want. Unless we extract the `PyDataFrame` from it, which is stupid because then we're building a `DataFrame` object twice.

That is why I created `construction.py` to have construction logic that builds just the `PyDataFrame`, which we can use in the `__init__` to finish building our `DataFrame. Makes perfect sense to me.

And now you want to specify a whole host of `DataFrame._from_x` methods to skip a cheap type check in the constructor. Doesn't sound very Pythonic to me.

Reply:
Ok, I understand  your concerns. To me the branching of generic to specialized makes sense. If this can be done with the specialized `pydataframe, pyseries` creator functions it also makes sense to me. So maybe the `Series._from_x` should call a `_pyseries_from_x`? 

Given that the function calls are more expensive than the branches and that the branches still call specialized pyseries constructors, I agree. 👍

Reply:
If we do want specialized `Series._from_x` methods that skip the constructor, I think this would be nicest, utilizing the constructor logic from `construction.py`:

```python
    @classmethod
    def _from_arrow(cls, name: str, values: pa.Array) -> "Series":
        return cls._from_pyseries(arrow_to_pyseries(name, values))
```

Reply:
Yes. That is exactly fixes the gripe I had with that. 💯

Reply:
Awesome! I think we arrived at the best solution here together :) Going to implement this.

Reply:
@ritchie46 Fixed! I think the PR should be OK now?